File: typescript/src/example-lakebase.ts
/**
 * LakebaseExampleStore — pgvector-backed `ExampleStore` over Databricks Lakebase.
 *
 * Sibling to `LakebaseMemoryStore`. Examples are scoped by `agentId` and
 * partitioned by `intent`. Embedding indexes the `input` field (the
 * discriminator queries are compared against) — `output` is stored verbatim
 * but not embedded.
 *
 * Schema (created on first use when `autoCreate=true`):
 *
 * ```sql
 * CREATE EXTENSION IF NOT EXISTS vector;
 * CREATE TABLE IF NOT EXISTS {table_name} (
 *   id          TEXT PRIMARY KEY,
 *   agent_id    TEXT NOT NULL,
 *   intent      TEXT NOT NULL DEFAULT 'default',
 *   input       TEXT NOT NULL,
 *   output      TEXT NOT NULL,
 *   score       REAL,
 *   tags        TEXT[] NOT NULL DEFAULT '{}',
 *   embedding   vector({embedding_dim}),
 *   metadata    JSONB NOT NULL DEFAULT '{}',
 *   created_at  DOUBLE PRECISION,
 *   updated_at  DOUBLE PRECISION
 * );
 * ```
 */

import { z } from 'zod';
import {
  applyExamplePatch,
  isoNow,
  materializeExample,
  newExampleId,
  type Example,
  type ExampleFilter,
  type ExampleInput,
  type ExamplePatch,
  type ExampleResult,
  type ExampleStore,
  type FindSimilarOptions,
} from './example.js';
import type { EmbeddingFn } from './memory.js';
import { tagsLiteral, vectorLiteral } from './memory-lakebase.js';
import type {
  PgExecExecutor,
  PgQueryExecutor,
} from './memory-lakebase.js';

// Re-export the executor types so callers using only example-lakebase have
// a single import target.
export type { PgExecExecutor, PgQueryExecutor } from './memory-lakebase.js';

// ---------------------------------------------------------------------------
// Options
// ---------------------------------------------------------------------------

const optionsSchema = z.object({
  tableName: z.string().min(1).optional(),
  autoCreate: z.boolean().optional(),
  ensureExtension: z.boolean().optional(),
  embeddingDim: z.number().int().positive(),
});

export interface LakebaseExampleStoreOptions {
  /** Row-returning executor for parameterized SELECT. */
  querySql: PgQueryExecutor;
  /** Side-effecting executor for parameterized DDL / DML. */
  execSql: PgExecExecutor;
  /** Required: embedding function. pgvector schema is fixed-width. */
  embeddingFn: EmbeddingFn;
  /** Required: dimensionality the embedder produces. Must match the `vector(N)` column. */
  embeddingDim: number;
  /** Table name (may include schema prefix). Defaults to `"apx_examples"`. */
  tableName?: string;
  /** Run `CREATE TABLE IF NOT EXISTS` and index DDL on first use. Defaults to true. */
  autoCreate?: boolean;
  /** Issue `CREATE EXTENSION IF NOT EXISTS vector` alongside the table DDL. Defaults to true. */
  ensureExtension?: boolean;
}

// ---------------------------------------------------------------------------
// Helpers (parallel to memory-lakebase, intentionally duplicated to avoid
// reaching into another module's private API)
// ---------------------------------------------------------------------------

function parseVectorLiteral(raw: unknown): number[] | undefined {
  if (raw === null || raw === undefined) return undefined;
  if (Array.isArray(raw)) return raw.map((v) => Number(v));
  let s = String(raw).trim();
  if (s.length === 0) return undefined;
  if (s.startsWith('"') && s.endsWith('"')) s = s.slice(1, -1);
  if (s.startsWith('[') && s.endsWith(']')) s = s.slice(1, -1);
  if (s.length === 0) return [];
  return s.split(',').map((p) => Number(p.trim()));
}

function parseTagsLiteral(raw: unknown): string[] {
  if (raw === null || raw === undefined) return [];
  if (Array.isArray(raw)) return raw.map((v) => String(v));
  let s = String(raw).trim();
  if (s === '{}' || s.length === 0) return [];
  if (s.startsWith('{') && s.endsWith('}')) s = s.slice(1, -1);
  if (s.length === 0) return [];
  const out: string[] = [];
  let cur = '';
  let inQuote = false;
  for (let i = 0; i < s.length; i++) {
    const ch = s[i]!;
    if (ch === '\\' && i + 1 < s.length) {
      cur += s[i + 1];
      i += 1;
      continue;
    }
    if (ch === '"') {
      inQuote = !inQuote;
      continue;
    }
    if (ch === ',' && !inQuote) {
      out.push(cur);
      cur = '';
      continue;
    }
    cur += ch;
  }
  out.push(cur);
  return out;
}

function parseMetadata(raw: unknown): Record<string, unknown> {
  if (raw === null || raw === undefined) return {};
  if (typeof raw === 'object') return raw as Record<string, unknown>;
  try {
    const v = JSON.parse(String(raw));
    return v && typeof v === 'object' ? (v as Record<string, unknown>) : {};
  } catch {
    return {};
  }
}

function epochToIso(seconds: unknown): string {
  const n = Number(seconds);
  if (!Number.isFinite(n) || n <= 0) return new Date(0).toISOString();
  return new Date(n * 1000).toISOString();
}

function isoToEpoch(iso: string | undefined): number {
  if (!iso) return Date.now() / 1000;
  const ms = Date.parse(iso);
  return Number.isFinite(ms) ? ms / 1000 : Date.now() / 1000;
}

function rowToExample(row: Record<string, unknown>): Example {
  const embedding = parseVectorLiteral(row.embedding);
  return Object.freeze({
    id: String(row.id),
    agentId: String(row.agent_id),
    intent: String(row.intent ?? 'default'),
    input: String(row.input ?? ''),
    output: String(row.output ?? ''),
    score:
      row.score === null || row.score === undefined ? undefined : Number(row.score),
    tags: Object.freeze(parseTagsLiteral(row.tags)),
    embedding: embedding ? Object.freeze([...embedding]) : undefined,
    metadata: Object.freeze(parseMetadata(row.metadata)),
    createdAt: epochToIso(row.created_at),
    updatedAt: epochToIso(row.updated_at),
  });
}

// ---------------------------------------------------------------------------
// LakebaseExampleStore
// ---------------------------------------------------------------------------

export class LakebaseExampleStore implements ExampleStore {
  readonly tableName: string;
  readonly embeddingDim: number;
  private readonly querySql: PgQueryExecutor;
  private readonly execSql: PgExecExecutor;
  private readonly embeddingFn: EmbeddingFn;
  private autoCreate: boolean;
  private ensureExtension: boolean;
  private ensurePromise: Promise<void> | null = null;

  constructor(options: LakebaseExampleStoreOptions) {
    optionsSchema.parse({
      tableName: options.tableName,
      autoCreate: options.autoCreate,
      ensureExtension: options.ensureExtension,
      embeddingDim: options.embeddingDim,
    });
    if (typeof options.embeddingFn !== 'function') {
      throw new Error(
        'LakebaseExampleStore: `embeddingFn` is required — pgvector schema demands embeddings at write time.',
      );
    }
    this.tableName = options.tableName ?? 'apx_examples';
    this.embeddingDim = options.embeddingDim;
    this.querySql = options.querySql;
    this.execSql = options.execSql;
    this.embeddingFn = options.embeddingFn;
    this.autoCreate = options.autoCreate ?? true;
    this.ensureExtension = options.ensureExtension ?? true;
  }

  // -- schema bootstrap -------------------------------------------------

  private async ensureSchema(): Promise<void> {
    if (!this.autoCreate) return;
    if (this.ensurePromise) return this.ensurePromise;
    this.ensurePromise = this._runDdl();
    return this.ensurePromise;
  }

  private async _runDdl(): Promise<void> {
    try {
      if (this.ensureExtension) {
        await this.execSql('CREATE EXTENSION IF NOT EXISTS vector', []);
      }
      const ddl =
        `CREATE TABLE IF NOT EXISTS ${this.tableName} (` +
        `  id TEXT PRIMARY KEY,` +
        `  agent_id TEXT NOT NULL,` +
        `  intent TEXT NOT NULL DEFAULT 'default',` +
        `  input TEXT NOT NULL,` +
        `  output TEXT NOT NULL,` +
        `  score REAL,` +
        `  tags TEXT[] NOT NULL DEFAULT '{}',` +
        `  embedding vector(${this.embeddingDim}),` +
        `  metadata JSONB NOT NULL DEFAULT '{}',` +
        `  created_at DOUBLE PRECISION,` +
        `  updated_at DOUBLE PRECISION` +
        `)`;
      await this.execSql(ddl, []);
      await this.execSql(
        `CREATE INDEX IF NOT EXISTS ${this.tableName.replace(/\./g, '_')}_agent_idx ` +
          `ON ${this.tableName} (agent_id)`,
        [],
      );
      await this.execSql(
        `CREATE INDEX IF NOT EXISTS ${this.tableName.replace(/\./g, '_')}_embedding_idx ` +
          `ON ${this.tableName} USING ivfflat (embedding vector_cosine_ops) WITH (lists = 100)`,
        [],
      );
    } catch (e) {
      console.warn(
        `LakebaseExampleStore: DDL bootstrap on ${this.tableName} failed: ${String(e)} — ` +
          `subsequent reads/writes may fail until the schema exists.`,
      );
    }
  }

  // -- internal row write -----------------------------------------------

  private async insertRow(example: Example, embedding: readonly number[]): Promise<void> {
    const sql =
      `INSERT INTO ${this.tableName} ` +
      `(id, agent_id, intent, input, output, score, tags, embedding, metadata, created_at, updated_at) ` +
      `VALUES ($1, $2, $3, $4, $5, $6, $7::text[], $8::vector, $9::jsonb, $10, $11)`;
    await this.execSql(sql, [
      example.id,
      example.agentId,
      example.intent,
      example.input,
      example.output,
      example.score ?? null,
      tagsLiteral(example.tags),
      vectorLiteral(embedding),
      JSON.stringify(example.metadata ?? {}),
      isoToEpoch(example.createdAt),
      isoToEpoch(example.updatedAt),
    ]);
  }

  // -- ExampleStore protocol --------------------------------------------

  async add(input: ExampleInput): Promise<Example> {
    await this.ensureSchema();
    const [embedding] = await this.embeddingFn([input.input]);
    if (!embedding || embedding.length !== this.embeddingDim) {
      throw new Error(
        `LakebaseExampleStore.add: embedder returned ${embedding?.length ?? 0} dims, ` +
          `expected ${this.embeddingDim}.`,
      );
    }
    const example = materializeExample(
      { ...input, id: input.id ?? newExampleId() },
      embedding,
    );
    await this.insertRow(example, embedding);
    return example;
  }

  async addBatch(inputs: readonly ExampleInput[]): Promise<Example[]> {
    if (inputs.length === 0) return [];
    await this.ensureSchema();
    const embeddings = await this.embeddingFn(inputs.map((i) => i.input));
    if (embeddings.length !== inputs.length) {
      throw new Error(
        `LakebaseExampleStore.addBatch: embedder returned ${embeddings.length} vectors, ` +
          `expected ${inputs.length}.`,
      );
    }
    const valueRows: string[] = [];
    const params: unknown[] = [];
    const out: Example[] = [];
    let p = 1;
    for (let i = 0; i < inputs.length; i++) {
      const emb = embeddings[i]!;
      if (!emb || emb.length !== this.embeddingDim) {
        throw new Error(
          `LakebaseExampleStore.addBatch: row ${i} embedder returned ${emb?.length ?? 0} dims, ` +
            `expected ${this.embeddingDim}.`,
        );
      }
      const example = materializeExample(
        { ...inputs[i]!, id: inputs[i]!.id ?? newExampleId() },
        emb,
      );
      out.push(example);
      valueRows.push(
        `($${p}, $${p + 1}, $${p + 2}, $${p + 3}, $${p + 4}, $${p + 5}, ` +
          `$${p + 6}::text[], $${p + 7}::vector, $${p + 8}::jsonb, $${p + 9}, $${p + 10})`,
      );
      p += 11;
      params.push(
        example.id,
        example.agentId,
        example.intent,
        example.input,
        example.output,
        example.score ?? null,
        tagsLiteral(example.tags),
        vectorLiteral(emb),
        JSON.stringify(example.metadata ?? {}),
        isoToEpoch(example.createdAt),
        isoToEpoch(example.updatedAt),
      );
    }
    const sql =
      `INSERT INTO ${this.tableName} ` +
      `(id, agent_id, intent, input, output, score, tags, embedding, metadata, created_at, updated_at) ` +
      `VALUES ${valueRows.join(', ')}`;
    await this.execSql(sql, params);
    return out;
  }

  async get(id: string): Promise<Example | null> {
    await this.ensureSchema();
    const sql =
      `SELECT id, agent_id, intent, input, output, score, tags, embedding, metadata, ` +
      `created_at, updated_at FROM ${this.tableName} WHERE id = $1 LIMIT 1`;
    try {
      const rows = await this.querySql(sql, [id]);
      if (!rows || rows.length === 0) return null;
      return rowToExample(rows[0]!);
    } catch (e) {
      console.warn(`LakebaseExampleStore.get(${id}) failed: ${String(e)} — returning null.`);
      return null;
    }
  }

  async update(id: string, patch: ExamplePatch): Promise<Example | null> {
    await this.ensureSchema();
    const existing = await this.get(id);
    if (!existing) return null;
    let next = applyExamplePatch(existing, patch);

    let embedding: readonly number[] | undefined = existing.embedding;
    if (patch.input !== undefined) {
      const [emb] = await this.embeddingFn([patch.input]);
      if (!emb || emb.length !== this.embeddingDim) {
        throw new Error(
          `LakebaseExampleStore.update: embedder returned ${emb?.length ?? 0} dims, ` +
            `expected ${this.embeddingDim}.`,
        );
      }
      embedding = emb;
      next = Object.freeze({ ...next, embedding: Object.freeze([...emb]) });
    }

    const sql =
      `UPDATE ${this.tableName} SET ` +
      `agent_id = $1, intent = $2, input = $3, output = $4, score = $5, ` +
      `tags = $6::text[], embedding = $7::vector, metadata = $8::jsonb, ` +
      `created_at = $9, updated_at = $10 WHERE id = $11`;
    await this.execSql(sql, [
      next.agentId,
      next.intent,
      next.input,
      next.output,
      next.score ?? null,
      tagsLiteral(next.tags),
      embedding ? vectorLiteral(embedding) : vectorLiteral(new Array(this.embeddingDim).fill(0)),
      JSON.stringify(next.metadata ?? {}),
      isoToEpoch(next.createdAt),
      isoToEpoch(next.updatedAt),
      id,
    ]);
    return next;
  }

  async delete(id: string): Promise<boolean> {
    await this.ensureSchema();
    const existed = (await this.get(id)) !== null;
    if (!existed) return false;
    const sql = `DELETE FROM ${this.tableName} WHERE id = $1`;
    try {
      await this.execSql(sql, [id]);
      return true;
    } catch (e) {
      console.warn(`LakebaseExampleStore.delete(${id}) failed: ${String(e)}`);
      return false;
    }
  }

  async list(filter: ExampleFilter): Promise<Example[]> {
    await this.ensureSchema();
    const limit = filter.limit ?? 100;
    const sql =
      `SELECT id, agent_id, intent, input, output, score, tags, embedding, metadata, ` +
      `created_at, updated_at FROM ${this.tableName} ` +
      `WHERE agent_id = $1 ` +
      `AND ($2::text IS NULL OR intent = $2) ` +
      `AND ($3::text[] IS NULL OR tags && $3::text[]) ` +
      `AND ($4::real IS NULL OR (score IS NOT NULL AND score >= $4)) ` +
      `ORDER BY updated_at DESC LIMIT $5`;
    const params: unknown[] = [
      filter.agentId,
      filter.intent ?? null,
      filter.tags && filter.tags.length > 0 ? tagsLiteral(filter.tags) : null,
      filter.minScore ?? null,
      limit,
    ];
    try {
      const rows = await this.querySql(sql, params);
      return rows.map((r) => rowToExample(r));
    } catch (e) {
      console.warn(`LakebaseExampleStore.list failed: ${String(e)} — returning [].`);
      return [];
    }
  }

  async findSimilar(opts: FindSimilarOptions): Promise<ExampleResult[]> {
    await this.ensureSchema();
    const k = opts.k ?? 5;
    const [queryEmbedding] = await this.embeddingFn([opts.query]);
    if (!queryEmbedding || queryEmbedding.length !== this.embeddingDim) {
      throw new Error(
        `LakebaseExampleStore.findSimilar: embedder returned ${queryEmbedding?.length ?? 0} dims, ` +
          `expected ${this.embeddingDim}.`,
      );
    }
    const sql =
      `SELECT id, agent_id, intent, input, output, score, tags, embedding, metadata, ` +
      `created_at, updated_at, 1 - (embedding <=> $1::vector) AS similarity ` +
      `FROM ${this.tableName} ` +
      `WHERE agent_id = $2 ` +
      `AND ($3::text IS NULL OR intent = $3) ` +
      `AND ($4::text[] IS NULL OR tags && $4::text[]) ` +
      `AND ($5::real IS NULL OR (score IS NOT NULL AND score >= $5)) ` +
      `AND embedding IS NOT NULL ` +
      `ORDER BY embedding <=> $1::vector LIMIT $6`;
    const params: unknown[] = [
      vectorLiteral(queryEmbedding),
      opts.agentId,
      opts.intent ?? null,
      opts.tags && opts.tags.length > 0 ? tagsLiteral(opts.tags) : null,
      opts.minScore ?? null,
      k,
    ];
    try {
      const rows = await this.querySql(sql, params);
      return rows.map((r) => ({
        example: rowToExample(r),
        score: Number(r.similarity ?? 0),
      }));
    } catch (e) {
      console.warn(`LakebaseExampleStore.findSimilar failed: ${String(e)} — returning [].`);
      return [];
    }
  }

  // -- internal test hooks ----------------------------------------------

  /** @internal — Test helper to disable auto-create. */
  _disableAutoCreate(): void {
    this.autoCreate = false;
    this.ensurePromise = Promise.resolve();
  }
}

export { isoNow };

File: python/src/apx_agent/_example_lakebase.py
"""LakebaseExampleStore — Postgres-backed few-shot example store.

Sibling to :class:`LakebaseMemoryStore`. Persists :class:`Example` rows
with pgvector-backed semantic retrieval over the ``input`` field.

Schema (created on first use when ``auto_create=True``)::

    CREATE EXTENSION IF NOT EXISTS vector;
    CREATE TABLE IF NOT EXISTS {table} (
        id          TEXT PRIMARY KEY,
        agent_id    TEXT NOT NULL,
        intent      TEXT NOT NULL DEFAULT 'default',
        input       TEXT NOT NULL,
        output      TEXT NOT NULL,
        score       REAL,
        tags        TEXT[] NOT NULL DEFAULT '{}',
        embedding   vector({dim}),
        metadata    JSONB NOT NULL DEFAULT '{}',
        created_at  DOUBLE PRECISION,
        updated_at  DOUBLE PRECISION
    );
    CREATE INDEX IF NOT EXISTS {table}_agent_idx ON {table} (agent_id);
    CREATE INDEX IF NOT EXISTS {table}_embedding_idx ON {table}
        USING ivfflat (embedding vector_cosine_ops) WITH (lists = 100);

Requires the ``lakebase`` extra::

    pip install 'apx-agent[lakebase]'
"""

from __future__ import annotations

import json
import logging
from collections.abc import Mapping, Sequence
from typing import TYPE_CHECKING, Any

from ._example import (
    Example,
    ExampleFilter,
    ExampleResult,
    FindSimilarOptions,
    apply_example_patch,
    materialize_example,
)
from ._memory import EmbeddingFn
from ._memory_lakebase import (
    _SQLALCHEMY_MISSING,
    _epoch_to_iso,
    _iso_to_epoch,
    _require_sqlalchemy,
    _vector_literal,
)

if TYPE_CHECKING:
    from sqlalchemy.engine import Engine

logger = logging.getLogger(__name__)


# Re-exported for symmetry with the memory module's friendly error message.
_ = _SQLALCHEMY_MISSING


# ---------------------------------------------------------------------------
# Row → Example
# ---------------------------------------------------------------------------


def _row_to_example(row: Mapping[str, Any]) -> Example:
    """Convert a SQL row to an :class:`Example`.

    Handles all the DB-type → Python-type quirks: pgvector strings, JSONB
    coming back as a string from psycopg, TEXT[] arrays, epoch timestamps.
    """
    tags_raw = row.get("tags")
    if tags_raw is None:
        tags: tuple[str, ...] = ()
    elif isinstance(tags_raw, (list, tuple)):
        tags = tuple(str(t) for t in tags_raw)
    else:
        tags = (str(tags_raw),) if tags_raw else ()

    metadata_raw = row.get("metadata")
    if metadata_raw is None:
        metadata: dict[str, Any] = {}
    elif isinstance(metadata_raw, str):
        try:
            metadata = json.loads(metadata_raw)
        except json.JSONDecodeError:
            metadata = {}
    elif isinstance(metadata_raw, dict):
        metadata = dict(metadata_raw)
    else:
        metadata = {}

    embedding_raw = row.get("embedding")
    embedding: tuple[float, ...] | None
    if embedding_raw is None:
        embedding = None
    elif isinstance(embedding_raw, str):
        stripped = embedding_raw.strip().strip("[]")
        embedding = (
            ()
            if not stripped
            else tuple(float(v) for v in stripped.split(","))
        )
    elif isinstance(embedding_raw, (list, tuple)):
        embedding = tuple(float(v) for v in embedding_raw)
    else:
        embedding = None

    score_raw = row.get("score")
    return Example(
        id=str(row["id"]),
        agent_id=str(row["agent_id"]),
        intent=str(row.get("intent") or "default"),
        input=str(row.get("input") or ""),
        output=str(row.get("output") or ""),
        score=float(score_raw) if score_raw is not None else None,
        tags=tags,
        embedding=embedding,
        metadata=metadata,
        created_at=_epoch_to_iso(row.get("created_at")),
        updated_at=_epoch_to_iso(row.get("updated_at")),
    )


# ---------------------------------------------------------------------------
# LakebaseExampleStore
# ---------------------------------------------------------------------------


class LakebaseExampleStore:
    """:class:`ExampleStore` backed by a Lakebase (or any Postgres) instance.

    Uses pgvector for semantic retrieval — the ``input`` field's embedding
    is compared against the query embedding via ``embedding <=> :q::vector``
    (cosine distance); ``score = 1 - distance``.

    Args:
        engine: SQLAlchemy ``Engine`` connected to the Lakebase instance.
        embedding_fn: Required batched embedding function. Called on
            :meth:`add` (per ``input``), :meth:`add_batch`, and
            :meth:`find_similar` (per query). Constructor raises
            ``ValueError`` if omitted.
        embedding_dim: Required dimensionality of ``embedding_fn``'s output.
        table_name: Defaults to ``"apx_examples"``.
        auto_create: When ``True`` (default), runs the DDL block on first use.
        ensure_extension: When ``True`` (default), the DDL block also runs
            ``CREATE EXTENSION IF NOT EXISTS vector``.
    """

    def __init__(
        self,
        *,
        engine: "Engine",
        embedding_fn: EmbeddingFn,
        embedding_dim: int,
        table_name: str = "apx_examples",
        auto_create: bool = True,
        ensure_extension: bool = True,
    ) -> None:
        """Construct the store. Raises ``ValueError`` on missing required args."""
        if embedding_fn is None:
            raise ValueError("LakebaseExampleStore requires an `embedding_fn`.")
        if embedding_dim is None or int(embedding_dim) <= 0:
            raise ValueError(
                "LakebaseExampleStore requires a positive `embedding_dim`."
            )
        _require_sqlalchemy()
        self.engine = engine
        self._embedding_fn = embedding_fn
        self.embedding_dim = int(embedding_dim)
        self.table_name = table_name
        self._auto_create = auto_create
        self._ensure_extension = ensure_extension
        self._created = False

    # -- DDL ----------------------------------------------------------------

    def _ensure_schema(self) -> None:
        """Run the table + index DDL exactly once per instance."""
        if self._created or not self._auto_create:
            return
        self._created = True
        sa = _require_sqlalchemy()
        table = self.table_name
        statements: list[str] = []
        if self._ensure_extension:
            statements.append("CREATE EXTENSION IF NOT EXISTS vector")
        statements.append(
            f"CREATE TABLE IF NOT EXISTS {table} ("
            f"  id TEXT PRIMARY KEY,"
            f"  agent_id TEXT NOT NULL,"
            f"  intent TEXT NOT NULL DEFAULT 'default',"
            f"  input TEXT NOT NULL,"
            f"  output TEXT NOT NULL,"
            f"  score REAL,"
            f"  tags TEXT[] NOT NULL DEFAULT '{{}}',"
            f"  embedding vector({self.embedding_dim}),"
            f"  metadata JSONB NOT NULL DEFAULT '{{}}',"
            f"  created_at DOUBLE PRECISION,"
            f"  updated_at DOUBLE PRECISION"
            f")"
        )
        statements.append(
            f"CREATE INDEX IF NOT EXISTS {table}_agent_idx "
            f"ON {table} (agent_id)"
        )
        statements.append(
            f"CREATE INDEX IF NOT EXISTS {table}_embedding_idx "
            f"ON {table} USING ivfflat (embedding vector_cosine_ops) "
            f"WITH (lists = 100)"
        )
        try:
            with self.engine.begin() as conn:
                for stmt in statements:
                    conn.execute(sa.text(stmt))
        except Exception as e:
            logger.warning(
                "LakebaseExampleStore: DDL on %s failed: %s — "
                "subsequent reads/writes may fail until the table exists.",
                table, e,
            )

    # -- ExampleStore protocol ---------------------------------------------

    def add(self, example: Mapping[str, Any]) -> Example:
        """Insert one example; returns the materialized row."""
        self._ensure_schema()
        embedding = self._embedding_fn([str(example["input"])])[0]
        materialized = materialize_example(example, embedding=embedding)
        self._insert_rows([materialized])
        return materialized

    def add_batch(self, examples: Sequence[Mapping[str, Any]]) -> list[Example]:
        """Insert many examples in one round-trip."""
        if len(examples) == 0:
            return []
        self._ensure_schema()
        embeddings = self._embedding_fn([str(e["input"]) for e in examples])
        out = [
            materialize_example(raw, embedding=emb)
            for raw, emb in zip(examples, embeddings)
        ]
        self._insert_rows(out)
        return out

    def _insert_rows(self, rows: Sequence[Example]) -> None:
        """Execute the per-row UPSERT SQL for a batch of materialized rows."""
        sa = _require_sqlalchemy()
        sql = sa.text(
            f"INSERT INTO {self.table_name} "
            f"(id, agent_id, intent, input, output, score, tags, "
            f" embedding, metadata, created_at, updated_at) "
            f"VALUES (:id, :agent_id, :intent, :input, :output, :score, :tags, "
            f"        CAST(:embedding AS vector), CAST(:metadata AS jsonb), "
            f"        :created_at, :updated_at) "
            f"ON CONFLICT (id) DO UPDATE SET "
            f"  agent_id = EXCLUDED.agent_id, "
            f"  intent = EXCLUDED.intent, "
            f"  input = EXCLUDED.input, "
            f"  output = EXCLUDED.output, "
            f"  score = EXCLUDED.score, "
            f"  tags = EXCLUDED.tags, "
            f"  embedding = EXCLUDED.embedding, "
            f"  metadata = EXCLUDED.metadata, "
            f"  updated_at = EXCLUDED.updated_at"
        )
        with self.engine.begin() as conn:
            for ex in rows:
                conn.execute(sql, self._row_params(ex))

    def _row_params(self, ex: Example) -> dict[str, Any]:
        """Build the bind params for an insert/update of one :class:`Example`."""
        return {
            "id": ex.id,
            "agent_id": ex.agent_id,
            "intent": ex.intent,
            "input": ex.input,
            "output": ex.output,
            "score": ex.score,
            "tags": list(ex.tags),
            "embedding": _vector_literal(ex.embedding),
            "metadata": json.dumps(dict(ex.metadata), default=str),
            "created_at": _iso_to_epoch(ex.created_at),
            "updated_at": _iso_to_epoch(ex.updated_at),
        }

    def get(self, example_id: str) -> Example | None:
        """Return the example with ``example_id``, or ``None`` if missing."""
        self._ensure_schema()
        sa = _require_sqlalchemy()
        sql = sa.text(
            f"SELECT id, agent_id, intent, input, output, score, tags, "
            f" embedding, metadata, created_at, updated_at "
            f"FROM {self.table_name} WHERE id = :id LIMIT 1"
        )
        try:
            with self.engine.connect() as conn:
                row = conn.execute(sql, {"id": example_id}).mappings().first()
        except Exception as e:
            logger.warning(
                "LakebaseExampleStore.get(%s) failed: %s — returning None.",
                example_id, e,
            )
            return None
        if row is None:
            return None
        return _row_to_example(dict(row))

    def update(self, example_id: str, patch: Mapping[str, Any]) -> Example | None:
        """Apply ``patch`` and return the updated row, or ``None`` if missing."""
        existing = self.get(example_id)
        if existing is None:
            return None
        updated = apply_example_patch(existing, patch)
        if "input" in patch:
            new_emb = self._embedding_fn([str(patch["input"])])[0]
            updated = Example(
                id=updated.id,
                agent_id=updated.agent_id,
                intent=updated.intent,
                input=updated.input,
                output=updated.output,
                score=updated.score,
                tags=updated.tags,
                embedding=tuple(float(v) for v in new_emb),
                metadata=updated.metadata,
                created_at=updated.created_at,
                updated_at=updated.updated_at,
            )
        self._insert_rows([updated])
        return updated

    def delete(self, example_id: str) -> bool:
        """Delete the row; returns ``True`` on hit, ``False`` on miss."""
        self._ensure_schema()
        sa = _require_sqlalchemy()
        sql = sa.text(f"DELETE FROM {self.table_name} WHERE id = :id")
        try:
            with self.engine.begin() as conn:
                result = conn.execute(sql, {"id": example_id})
        except Exception as e:
            logger.warning(
                "LakebaseExampleStore.delete(%s) failed: %s", example_id, e
            )
            return False
        rowcount = getattr(result, "rowcount", -1)
        return rowcount is None or rowcount > 0

    def list(self, filter: ExampleFilter) -> list[Example]:
        """Return rows matching ``filter``, sorted by ``updated_at`` desc."""
        self._ensure_schema()
        sa = _require_sqlalchemy()
        where_clauses = ["agent_id = :agent_id"]
        params: dict[str, Any] = {
            "agent_id": filter.agent_id,
            "limit": filter.limit,
        }
        if filter.intent is not None:
            where_clauses.append("intent = :intent")
            params["intent"] = filter.intent
        if filter.tags:
            where_clauses.append("tags && CAST(:tags AS TEXT[])")
            params["tags"] = list(filter.tags)
        if filter.min_score is not None:
            # Rows with NULL score are excluded — mirrors InMemory behavior.
            where_clauses.append("score IS NOT NULL AND score >= :min_score")
            params["min_score"] = float(filter.min_score)
        sql = sa.text(
            f"SELECT id, agent_id, intent, input, output, score, tags, "
            f" embedding, metadata, created_at, updated_at "
            f"FROM {self.table_name} "
            f"WHERE {' AND '.join(where_clauses)} "
            f"ORDER BY updated_at DESC "
            f"LIMIT :limit"
        )
        # Let infra failures propagate: a swallowed error here would read as "no
        # examples" to prompt assembly and silently degrade answer quality.
        # Reserve [] for a genuinely empty result set. (matches the memory store, H21) (#380)
        with self.engine.connect() as conn:
            rows = conn.execute(sql, params).mappings().all()
        return [_row_to_example(dict(r)) for r in rows]

    def find_similar(self, opts: FindSimilarOptions) -> list[ExampleResult]:
        """pgvector retrieval: ``embedding <=> :q::vector`` = cosine distance.

        ``score = 1 - distance``; ordered ascending by distance.
        """
        self._ensure_schema()
        sa = _require_sqlalchemy()
        query_emb = self._embedding_fn([opts.query])[0]
        where_clauses = ["agent_id = :agent_id"]
        params: dict[str, Any] = {
            "agent_id": opts.agent_id,
            "query_vec": _vector_literal(query_emb),
            "k": opts.k,
        }
        if opts.intent is not None:
            where_clauses.append("intent = :intent")
            params["intent"] = opts.intent
        if opts.tags:
            where_clauses.append("tags && CAST(:tags AS TEXT[])")
            params["tags"] = list(opts.tags)
        if opts.min_score is not None:
            where_clauses.append("score IS NOT NULL AND score >= :min_score")
            params["min_score"] = float(opts.min_score)
        where_clauses.append("embedding IS NOT NULL")
        sql = sa.text(
            f"SELECT id, agent_id, intent, input, output, score, tags, "
            f" embedding, metadata, created_at, updated_at, "
            f" 1 - (embedding <=> CAST(:query_vec AS vector)) AS sim "
            f"FROM {self.table_name} "
            f"WHERE {' AND '.join(where_clauses)} "
            f"ORDER BY embedding <=> CAST(:query_vec AS vector) ASC "
            f"LIMIT :k"
        )
        # Let infra failures propagate (see .list) — a swallowed error would read
        # as "no examples" and silently degrade answer quality. (#380)
        with self.engine.connect() as conn:
            rows = conn.execute(sql, params).mappings().all()
        results: list[ExampleResult] = []
        for r in rows:
            d = dict(r)
            sim = float(d.pop("sim", 0.0) or 0.0)
            results.append(ExampleResult(example=_row_to_example(d), score=sim))
        return results


__all__ = ["LakebaseExampleStore"]
